enterprise/backend/src/metabase_enterprise/permission_debug/impl.clj

```diff
@@ -0,0 +1,235 @@
+                           permission-filter) :perm_grant] [:= :blocked.id :perm_grant.id]
+                         [:metabase_database :db] [:= :blocked.db_id :db.id]
+                         [:permissions_group :pg] [:= :perm_grant.group_id :pg.id]]
+                  :where [:and [:= :blocked.db_id (:database_id card)]
+                          [:not
+                           [:in :blocked.id (perms/visible-table-filter-select
+                                             :db_id
+                                             {:user-id user-id
+                                              :is-superuser? user-is-superuser?}
+                                             visible-filter)]]]})
+
+       (seq query-tables)
+       (t2/query {:select [[:db.name :db_name] :blocked.schema [:blocked.name :table_name] [:pg.name :group_name]]
+                  :from [[:metabase_table :blocked]]
+                  :join [[(perms/select-tables-and-groups-granting-perm
+                           {:user-id user-id
+                            :is-superuser? user-is-superuser?}
+                           permission-filter) :perm_grant] [:= :blocked.id :perm_grant.id]
+                         [:metabase_database :db] [:= :blocked.db_id :db.id]
+                         [:permissions_group :pg] [:= :perm_grant.group_id :pg.id]]
+                  :where [:and [:in :blocked.id query-tables]
+                          [:not
+                           [:in :blocked.id (perms/visible-table-filter-select
+                                             :id
+                                             {:user-id user-id
+                                              :is-superuser? user-is-superuser?}
+                                             visible-filter)]]]})
+       :else
+       nil)
+     (map (juxt (juxt :db_name :schema :table_name) :group_name))
+     (reduce #(update %1 (first %2) set/union (set [(second %2)])) {}))))
+
+(mu/defmethod debug-permissions :card/query :- ::DebuggerSchema
+  [{:keys [user-id model-id] :as debug-info} :- ::DebuggerArguments]
+  (merge-permission-check
+   (debug-permissions (assoc debug-info :permission-type :card/read))
+   (let [card-id (Integer/parseInt model-id)
+         card (t2/select-one :model/Card :id card-id)
+         blocked-by-group (check-table-permissions user-id card
+                                                   {:perms/view-data [:blocked :most]}
+                                                   {:perms/view-data :legacy-no-self-service})]
+     {:decision    (if (seq blocked-by-group) "denied" "allow")
+      :segment     #{}
+      :model-type  "question"
+      :model-id    model-id
+      :message     [(if (seq blocked-by-group)
+                      (tru "User does not have permission to query this card")
+                      (tru "User has permission to query this card"))]
+      :data        (if blocked-by-group (format-blocked blocked-by-group) {})
```

Comment:
this feels a bit awkward, the `data` here is a map of blocked tables to the group name that blocked them, but I'm not sure that feels natural? Like should it instead be a map like `{:blocked-tables {"PUBLIC.foo" ["All Users"]}}` or something that makes it a bit more clear what the data actually represents?